include/api/Metrics.h

```diff
@@ -289,10 +280,126 @@ class Metrics
 
   mutable std::mutex _mutex;
   LookupTable _lookups;
-  MetricBlobs _blobs;
+  BlobStorage _blobs;
   uint16_t _cur_blob = 0;
   uint16_t _cur_off  = 0;
 
+public:
+  // These are sort of factory classes, using the Metrics singleton for all storage etc.
+  class Gauge
+  {
+  public:
+    using self_type = Gauge;
+    using SpanType  = Metrics::SpanType;
+
+    class AtomicType : public Metrics::AtomicType
+    {
+    };
+
+    static Metrics::IdType
+    create(const std::string_view name)
+    {
+      auto &instance = Metrics::instance();
+
+      return instance._create(name);
+    }
+
+    static AtomicType *
+    createPtr(const std::string_view name)
+    {
+      auto &instance = Metrics::instance();
+
+      return reinterpret_cast<AtomicType *>(instance.lookup(instance._create(name)));
+    }
+
+    static Metrics::Gauge::SpanType
+    createSpan(size_t size, IdType *id = nullptr)
+    {
+      auto &instance = Metrics::instance();
+
+      return instance._createSpan(size, id);
+    }
+
+    static void
+    increment(AtomicType *metric, uint64_t val = 1)
+    {
+      ink_assert(metric);
+      metric->_value.fetch_add(val, MEMORY_ORDER);
+    }
+
+    static void
+    decrement(AtomicType *metric, uint64_t val = 1)
+    {
+      ink_assert(metric);
+      metric->_value.fetch_sub(val, MEMORY_ORDER);
+    }
+
+    static int64_t
+    load(AtomicType *metric)
+    {
+      ink_assert(metric);
+      return metric->_value.load();
+    }
+
+    static void
+    store(AtomicType *metric, int64_t val)
+    {
+      ink_assert(metric);
+      return metric->_value.store(val);
+    }
+
+  }; // class Gauge
+
+  class Counter
+  {
+  public:
+    using self_type = Counter;
+    using SpanType  = Metrics::SpanType;
+
+    class AtomicType : public Metrics::AtomicType
+    {
+    };
+
+    static Metrics::IdType
+    create(const std::string_view name)
+    {
+      auto &instance = Metrics::instance();
+
+      return instance._create(name);
+    }
+
+    static AtomicType *
+    createPtr(const std::string_view name)
+    {
+      auto &instance = Metrics::instance();
+
+      return reinterpret_cast<AtomicType *>(instance.lookup(instance._create(name)));
+    }
+
+    static Metrics::Counter::SpanType
+    createSpan(size_t size, IdType *id = nullptr)
+    {
+      auto &instance = Metrics::instance();
+
+      return instance._createSpan(size, id);
+    }
+
+    static void
+    increment(AtomicType *metric, uint64_t val = 1)
+    {
+      ink_assert(metric);
+      metric->_value.fetch_add(val, MEMORY_ORDER);
+    }
+
+    static int64_t
+    load(AtomicType *metric)
```

Comment:
This can't be const?

Reply:
Maybe? I can try.

Reply:
Ah yeh, I remember now, it can't. I blame the Danes.

```
/Users/leif/apache/trafficserver/include/api/Metrics.h:395:30: error: static member function cannot have 'const' qualifier
    load(AtomicType *metric) const
```

Reply:
Ok, fixed, and uploaded with a new commit for your review suggestions (will add more as if we identify more things that needs fixing).